astrobee/survey_manager/survey_planner/tools/survey_planner.py

```diff
@@ -0,0 +1,958 @@
+        string. The default implementation always returns an empty string. Derived classes can
+        override as needed.
+        """
+        return ""
+
+    def is_valid(self, sim_state: "SimState") -> bool:
+        """
+        Return True if the action can be executed in `sim_state`.
+        """
+        return self.invalid_reason(sim_state) == ""
+
+    def get_pddl_name(self) -> str:
+        "Return the name of this action type in the PDDL model."
+        return self.__class__.__name__.replace("Action", "").lower()
+
+    @abstractmethod
+    def get_pddl_args(self) -> List[Any]:
+        "Return the arguments of this action type in the PDDL model."
+
+    def __repr__(self):
+        args_str = " ".join((str(arg) for arg in self.get_pddl_args()))
+        return f"({self.get_pddl_name()} {args_str})"
+
+    def start(self, sim_state: "SimState") -> None:
+        """
+        Modify `sim_state` as needed at the beginning of action execution. Derived classes that
+        need to extend start() should typically invoke the parent method.
+        """
+        sim_state.trace.append((sim_state.elapsed_time, self, self.get_duration()))
+        robot_state = sim_state.robot_states[self.robot]
+        robot_state.action = self
+        # It seems to be a PDDL convention to separate events by an epsilon time difference
+        # to prevent ambiguity about ordering.
+        sim_state.elapsed_time += 0.001
+
+    def end(self, sim_state: "SimState") -> None:
+        """
+        Modify `sim_state` as needed at the end of action execution. Derived classes that need to
+        extend end() should typically invoke the parent method.
+        """
+        robot_state = sim_state.robot_states[self.robot]
+        robot_state.action = None
+        # It seems to be a PDDL convention to separate events by an epsilon time difference
+        # to prevent ambiguity about ordering.
+        sim_state.elapsed_time += 0.001
+
+    def apply(self, sim_state: "SimState") -> None:
+        """
+        Apply the action, modifying `sim_state`.
+        """
+        end_time = sim_state.elapsed_time + self.get_duration()
+        self.start(sim_state)
+        heapq.heappush(sim_state.events, (end_time, self.end))
+
+
+class RobotState:
+    "Class representing the state of a robot."
+
+    def __init__(
+        self, pos: LocationName, action: Optional[Action], reserved: List[LocationName]
+    ):
+        """
+        :param pos: The current location of the robot.
+        :param action: The action the robot is currently executing (or None if it is idle).
+        :param reserved: A list of locations the robot currently has reserved. (Places the robot
+            is expected to move through given the action it is currently executing.)
+        """
+        self.pos = pos
+        self.action = action
+        self.reserved = reserved
+
+    def get_dump_dict(self) -> Dict[str, Any]:
```

Comment:
The overview of the dump approach here is to first recursively construct a nested object of built-in `dict` and `list` types, then serialize that to an output format like YAML. I find the `get_dump_dict()` methods to be a bit simpler than the corresponding `__str__()` methods would be, and the resulting hierarchical dump output is always "well-formed" at some level (balanced parens, proper indentation, etc.).

It would be natural to also define a generic `__str__()` method that uses `get_dump_dict()` and does the dump in every class, but I didn't actually need most of them.